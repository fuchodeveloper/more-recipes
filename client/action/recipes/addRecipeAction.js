import { batchActions } from 'redux-batched-actions';
import axios from 'axios';
import alertify from 'alertify.js';
import { ADD_RECIPE, ADD_RECIPE_ERROR } from '../types';
import { setFetching, unsetFetching } from '../fetching';

/**
 * @description create a recipe
 *
 * @export addRecipeActionCreator
 *
 * @param {Object} recipe
 *
 * @returns {Object} recipe
 */
const addRecipeActionCreator = recipe => ({
  type: ADD_RECIPE,
  recipe
});

/**
 * @description create a recipe error
 *
 * @export addRecipeActionError
 *
 * @param {Object} error
 *
 * @returns {Object} error
 */
const addRecipeActionError = error => ({
  type: ADD_RECIPE_ERROR,
  error
});

const addRecipeAction = recipe => (dispatch) => {
  dispatch(setFetching());
  return axios.post('/api/v1/recipes', recipe)
    .then((response) => {
      alertify.delay(900);
      alertify.logPosition('bottom right');
      alertify.success(response.data.message);
      dispatch(batchActions([
        dispatch(addRecipeActionCreator(response.data.recipe)),
        unsetFetching()
      ]));
    })
    .catch((error) => {
      dispatch([
        dispatch(addRecipeActionError(error.response.data.error)),
        unsetFetching()
      ]);
    });
};

export default addRecipeAction;
