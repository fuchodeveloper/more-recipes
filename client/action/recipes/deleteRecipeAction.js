import axios from 'axios';
import alertify from 'alertify.js';
import { DELETE_RECEIPE, DELETE_RECEIPE_FAIL } from '../types';

export const deleteRecipeActionError = error => ({
  type: DELETE_RECEIPE_FAIL,
  error
});

export const deleteRecipeActionCreator = recipes => ({
  type: DELETE_RECEIPE,
  recipes
});

const deleteRecipeAction = recipeId => dispatch =>
  axios.delete(`/api/v1/recipes/${recipeId}`)
    .then((response) => {
      alertify.delay(900);
      alertify.logPosition('bottom right');
      alertify.success(response.data.message);
      dispatch(deleteRecipeActionCreator(response.data.recipes));
    })
    .catch((error) => {
      alertify.delay(900);
      alertify.logPosition('bottom right');
      alertify.success(error.data.message);
      dispatch(deleteRecipeActionError(error.data.message));
    });

export default deleteRecipeAction;
