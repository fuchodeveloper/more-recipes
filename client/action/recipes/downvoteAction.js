import axios from 'axios';
import alertify from 'alertify.js';
import { GET_DOWNVOTE } from '../types';

/**
 * Recipe doenvote action creator
 * @param {recipe} recipe
 * @returns {recipe} recipe
 */
export const downvoteRecipeActionCreator = recipe => ({
  type: GET_DOWNVOTE,
  recipe
});

/**
 * Authenticated user can downvote a recipe
 *
 * @export func
 * @param {integer} param
 * @returns {object} object
 */
const downvoteRecipe = param =>
  dispatch => axios.post(`/api/v1/recipes/${param}/downvote`)
    .then((response) => {
      alertify.delay(900);
      alertify.logPosition('bottom right');
      alertify.success(response.data.message);
      return dispatch(downvoteRecipeActionCreator(response.data.recipe));
    })
    .catch((error) => {
      alertify.logPosition('bottom right');
      alertify.error(error.message);
    });


export default downvoteRecipe;
