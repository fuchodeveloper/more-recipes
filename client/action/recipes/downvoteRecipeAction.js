import axios from 'axios';
import alertify from 'alertify.js';
import { DOWNVOTE_RECIPE, DOWNVOTE_RECIPE_ERROR } from '../types';
import networkError from '../networkError';

/**
 * @description Recipe downvote action creator
 *
 * @param {recipe} recipe recipe parameter
 *
 * @returns {recipe} recipe downvoted
 */
export const downvoteRecipeActionCreator = recipe => ({
  type: DOWNVOTE_RECIPE,
  recipe
});
/**
 * @description Recipe downvote error
 *
 * @param {Object} error error parameter
 *
 * @returns {Object} error returns the error object
 *
 */
const downvoteRecipeActionError = error => ({
  type: DOWNVOTE_RECIPE_ERROR,
  error
});

/**
 * @description Authenticated user can downvote a recipe
 *
 * @export downvoteRecipe
 *
 * @param {Number} id
 *
 * @returns {object} downvoted recipe object
 */
const downvoteRecipeAction = id =>
  dispatch => axios.post(`/api/v1/recipes/${id}/downvote`)
    .then((response) => {
      alertify.delay(900);
      alertify.logPosition('bottom right');
      alertify.success(response.data.recipe.message);
      return dispatch(downvoteRecipeActionCreator(response.data.recipe));
    })
    .catch((error) => {
      if (!error.response) {
        return networkError(error);
      }
      alertify.delay(1000);
      alertify.logPosition('bottom right');
      alertify.error('Please login to downvote recipe');
      dispatch(downvoteRecipeActionError(error.response.data));
    });


export default downvoteRecipeAction;
