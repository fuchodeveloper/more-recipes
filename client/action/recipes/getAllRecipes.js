import axios from 'axios';
import { batchActions } from 'redux-batched-actions';
import { GET_ALL_RECIPES, GET_PAGE_DETAILS } from '../types';
import { setFetching, unsetFetching } from '../fetching';

export const receiveRecipes = recipes => ({
  type: GET_ALL_RECIPES,
  recipes
});

export const recipePageDetails = pageCount => ({
  type: GET_PAGE_DETAILS,
  pageCount
});

const getAllRecipes = page => (dispatch) => {
  dispatch(setFetching());
  axios.get(`/api/v1/recipes?page=${page}`)
    .then((response) => {
      dispatch(batchActions([
        receiveRecipes(response.data.recipes),
        recipePageDetails(response.data.pageCount),
        unsetFetching(),
      ]));
    })
    .catch((error) => {
      // dispatch(batchActions([

      // ]));
      console.log(error.message);
    });
};

export default getAllRecipes;
