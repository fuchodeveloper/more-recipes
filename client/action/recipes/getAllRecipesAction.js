import axios from 'axios';
import { batchActions } from 'redux-batched-actions';
import
{ GET_ALL_RECIPES, GET_PAGE_DETAILS, GET_ALL_RECIPES_ERROR } from '../types';
import { setFetching, unsetFetching } from '../fetching';

/**
 * @description get all recipes action creator
 *
 * @param {Object} recipes recipes object parameter
 *
 * @returns {Object} recipes returns all recipes object
 */
const receiveRecipes = recipes => ({
  type: GET_ALL_RECIPES,
  recipes
});

/**
 * @description get all recipes error action
 *
 * @param {Object} error recipes error object
 *
 * @returns {Object} error returns get all recipes error object
 */
const receiveRecipesError = error => ({
  type: GET_ALL_RECIPES_ERROR,
  error
});

/**
 * @description handles page detail
 *
 * @param {Number} pageCount recipe page count parameter
 *
 * @returns {Number} returns recipes page Count
 */
const recipePageCount = pageCount => ({
  type: GET_PAGE_DETAILS,
  pageCount
});

const getAllRecipesAction = page => (dispatch) => {
  dispatch(setFetching());
  return axios.get(`/api/v1/recipes?page=${page}`)
    .then((response) => {
      dispatch(batchActions([
        dispatch(receiveRecipes(response.data.recipes)),
        recipePageCount(response.data.pageCount),
        unsetFetching(),
      ]));
    })
    .catch((serverError) => {
      dispatch(receiveRecipesError(serverError.recipes.data.error));
    });
};

export default getAllRecipesAction;
