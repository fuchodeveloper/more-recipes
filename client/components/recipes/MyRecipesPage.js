import React from 'react';
import axios from 'axios';
import Header from '../../components/navigation/Header';
import Footer from '../../components/navigation/Footer';
import MyRecipesDetail from './MyRecipesDetail';

class MyRecipesPage extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      isLoading: true,
      details: {},
      errors: {},
      favoriteCount: 0,
      cloudinaryRecipeImage: ''
    }; // Initialize the state
  }

  /**
   *  GET all recipes for authenticated user using API endpoint
   */
  componentWillMount() {
    setTimeout(() => this.setState({ isLoading: false }), 1000);
    axios.get('/api/v1/recipes/get_all_for_user')
    .then((recipe) => {
      console.log(recipe.data);
      this.setState({ details: recipe.data })
    })
    .catch((error) => {
      console.log(error.data);
      this.setState({ errors: error.response })
    })
  }

  render() {
    const { isLoading, favoriteCount } = this.state;
    
    /**
     * Display loading text while fetching the recipes
     */
    if (isLoading) {
      return (
        <h2 className="text-center">Loading...</h2>
      );
    }

    return (
      <div>
        {/* Display page Header */}
        <Header />

        <div className="container margin-top-70">

          <div>
              <h1 className="text-center p-4 center-hero-title">My Recipes</h1>
          </div>

          <div className="margin-top-50 margin-bottom-50"/>

          {/* <div className="card-deck mb-5"> */}
          <div className="row">
          {Object
            .keys(this.state.details.recipes)
            .map(key => <MyRecipesDetail key={key} details={this.state.details.recipes[key]} />)  
          }

          </div>
{/* 
          <div className="card-deck mb-5">
              <div className="card mt-1">
                  <img className="card-img-top" src="./img/noodles.jpg" alt="Card image cap" />
                  <div className="card-body">
                      <h4 className="card-title favorite-recipe-title"><strong>Party Jollof Rice</strong></h4>
                      <p className="card-text recipe-font-family">This is a with supporting text below as a natural lead-in to additional content. This content is a little bit longer...</p>
                      <div className="card-footer custom-card-footer-bg">
                          <p className="card-text"><small className="text-muted">30 views . 20 votes . 42 reviews</small></p>
                          <a href="#" className="btn btn-primary btn-primary-color">View Recipe</a>
                      </div>
                  </div>
              </div>
              <div className="card mt-1">
                  <img className="card-img-top" src="./img/noodles.jpg" alt="Card image cap" />
                  <div className="card-body">
                      <h4 className="card-title favorite-recipe-title"><strong>Party Jollof Rice</strong></h4>
                      <p className="card-text recipe-font-family">This is a with supporting text below as a natural lead-in to additional content. This content is a little bit longer...</p>
                      <div className="card-footer custom-card-footer-bg">
                          <p className="card-text"><small className="text-muted">30 views . 20 votes . 42 reviews</small></p>
                          <a href="#" className="btn btn-primary btn-primary-color">View Recipe</a>
                      </div>
                  </div>
              </div>

              <div className="card mt-1">
                  <img className="card-img-top" src="./img/noodles.jpg" alt="Card image cap"/>
                  <div className="card-body">
                      <h4 className="card-title favorite-recipe-title"><strong>Party Jollof Rice</strong></h4>
                      <p className="card-text recipe-font-family">This is a with supporting text below as a natural lead-in to additional content. This content is a little bit longer...</p>
                      <div className="card-footer custom-card-footer-bg">
                          <p className="card-text"><small className="text-muted">30 views . 20 votes . 42 reviews</small></p>
                          <a href="#" className="btn btn-primary btn-primary-color">View Recipe</a>
                      </div>
                  </div>
              </div>

          </div> */}

          {/* Bottom Navigation */}
          <nav aria-label="Page navigation example">
              <ul className="pagination justify-content-center">
                  <li className="page-item disabled">
                      <a className="page-link" href="#" tabIndex="-1">Previous</a>
                  </li>
                  <li className="page-item"><a className="page-link" href="#">1</a></li>
                  <li className="page-item"><a className="page-link" href="#">2</a></li>
                  <li className="page-item"><a className="page-link" href="#">3</a></li>
                  <li className="page-item">
                      <a className="page-link" href="#">Next</a>
                  </li>
              </ul>
          </nav>


      </div>

        <div className="clearfix m-5"/>

        {/* Display footer */}
         <Footer />
      </div>
    );
  }
}

export default MyRecipesPage;
