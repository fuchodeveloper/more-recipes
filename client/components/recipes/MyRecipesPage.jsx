import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import alertify from 'alertify.js';
import ReactPaginate from 'react-paginate';
import { BounceLoader } from 'react-spinners';
import MyRecipesDetail from './MyRecipesDetail';
import myRecipesAction from '../../action/recipes/myRecipesActions';
import deleteRecipeAction from '../../action/recipes/deleteRecipeAction';

/**
 *
 *
 * @class MyRecipesPage
 *
 * @extends {React.Component}
 */
export class MyRecipesPage extends React.Component {
  /**
   * @description Creates an instance of MyRecipesPage.
   *
   * @param {Object} props
   *
   * @memberof MyRecipesPage
   */
  constructor(props) {
    super(props);
    this.state = {
      recipes: '',
      pageCount: ''
    }; // Initialize the state

    this.onPageChange = this.onPageChange.bind(this);
    this.onConfirm = this.onConfirm.bind(this);
  }


  /**
   * @description GET all recipes for authenticated user using API endpoint
   *
   * @memberof MyRecipesPage
   *
   * @returns {undefined}
   */
  componentDidMount() {
    this.props.myRecipesAction(this.state.pageCount);
  }
  /**
 * @description Component lifecycle method
 *
 * @param {Object} nextProps
 *
 * @memberof MyRecipesPage
 *
 * @returns {undefined}
 */
  componentWillReceiveProps(nextProps) {
    const { recipes } = nextProps;
    const { pageCount } = nextProps;
    this.setState({ recipes, pageCount });
  }


  /**
 * function to handle page number change
 *
 * @param {Number} current
 *
 * @memberof MyRecipesPage
 *
 * @returns {undefined}
 */
  onPageChange(current) {
    current.selected += 1;
    this.props.myRecipesAction(current.selected);
  }

  /**
 * @description confirm recipe deletion
 *
 * @param {Object} event
 *
 * @memberof MyRecipesPage
 *
 * @returns {undefined}
 */
  onConfirm(event) {
    event.preventDefault();
    const id = event.target.getAttribute('data-id');
    alertify.confirm('Do you really want to delete this recipe', () => {
      // user clicked "ok"
      this.props.deleteRecipeAction(id);
    }, () => {
      // user clicked "cancel"
    });
  }

  /**
 * @description renders component to DOM
 *
 * @memberof MyRecipesPage
 *
 *  @returns {JSX} JSX representation of component
 */
  render() {
    const { isFetching } = this.props;

    if (isFetching) {
      return (
        <h2 className="text-center">
          <span className="mt-5 d-inline-block">
            <BounceLoader
              color="#FF7055"
              loading={isFetching}
            />
          </span>
        </h2>
      );
    }


    if (this.state.recipes.length > 0) {
      return (
        <div>

          <div className="container margin-top-70">

            <div>
              <h1 className="text-center p-4 center-hero-title">My Recipes</h1>
            </div>

            <div className="margin-top-50 margin-bottom-50" />

            <div className="row">
              {Object
              .keys(this.state.recipes)
              .map(key => (<MyRecipesDetail
                key={key}
                details={this.state.recipes[key]}
                onConfirm={this.onConfirm}
              />))
            }

            </div>

            <div className="row">
              <ReactPaginate
                pageCount={parseInt(this.state.pageCount, 10)}
                pageRangeDisplayed={5}
                marginPagesDisplayed={3}
                previousLabel="Previous"
                nextLabel="Next"
                breakClassName="text-center"
                initialPage={0}
                containerClassName="container pagination justify-content-center"
                pageClassName="page-item"
                pageLinkClassName="page-link"
                activeClassName="page-item active"
                previousClassName="page-item"
                nextClassName="page-item"
                nextLinkClassName="page-link"
                previousLinkClassName="page-link"
                onPageChange={this.onPageChange}
              />
            </div>

          </div>

          <div className="clearfix m-5" />
        </div>
      );
    }

    return (
      <div>

        <div className="container margin-top-70">

          <div>
            <h1 className="text-center p-4 center-hero-title">My Recipes</h1>
          </div>

          <div className="margin-top-50 margin-bottom-50" />

          <div>
            <h2 className="text-center"><i className="recipe-title">
            You have not created any recipes
                                        </i>
            </h2>
            <div className="text-center">
              <Link to="/add" className="btn btn-primary btn-primary-color">
              Create a recipe
              </Link>
            </div>
          </div>
          <div className="clearfix mt-4" />

          <div className="clearfix m-5" />

        </div>
      </div>
    );
  }
}

MyRecipesPage.defaultProps = {
  recipes: {}
};

MyRecipesPage.propTypes = {
  myRecipesAction: PropTypes.func.isRequired,
  deleteRecipeAction: PropTypes.func.isRequired,
  recipes: PropTypes.shape({})
};

/**
 * @description function mapStateToProps
 *
 * @param {Object} state
 *
 * @returns {Object} recipes
 */
const mapStateToProps = state => ({
  recipes: state.recipesReducer.recipes,
  pageCount: state.recipesReducer.pageCount
});

/**
 * @description dispatch myRecipesActionProps
 *
 * @param {dispatch} dispatch
 *
 * @returns {dispatch} - dispatch myRecipesAction and deleteRecipeAction
 */
const mapDispatchToProps = dispatch => ({
  myRecipesAction: pageCount => dispatch(myRecipesAction(pageCount)),
  deleteRecipeAction: recipeId => dispatch(deleteRecipeAction(recipeId))
});

export default connect(mapStateToProps, mapDispatchToProps)(MyRecipesPage);
