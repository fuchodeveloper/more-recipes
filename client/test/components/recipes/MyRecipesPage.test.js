/* eslint-disable max-len */
import React from 'react';
import { shallow } from 'enzyme';
import expect from 'expect';
import sinon from 'sinon';
import { MyRecipesPage } from '../../../components/recipes/MyRecipesPage';

let props;

const setup = () => {
  props = {
    recipes: {},
    pageCount: '',
    myRecipesAction: jest.fn(),
    deleteRecipeAction: jest.fn()
  };
  return shallow(<MyRecipesPage {...props} />);
};


describe('MyRecipesPage component', () => {
  const page = {
    selected: 1
  };

  describe('MyRecipesPage component snapshot', () => {
    it('should render with right amount of elements', () => {
      const { wrapper } = setup();
      expect(wrapper).toMatchSnapshot();
    });
  });

  it('should call the onPageChange handler', () => {
    const spy = sinon.spy(MyRecipesPage.prototype, 'onPageChange');
    shallow(<MyRecipesPage {...props} onPageChange={spy} />)
      .instance().onPageChange(page);
  });

  it('component should receive props', () => {
    const wrapper = setup();
    const action = wrapper.instance();
    const spy = sinon.spy(
      MyRecipesPage.prototype,
      'componentWillReceiveProps'
    );
    const nextProps = {
      recipes: [
        {
          id: 1,
          userId: 1,
          name: 'jollof rice and beans',
          favoriteCount: 0,
          ingredients: 'rice and beans',
          direction: 'proin eget tortor risus. nulla quis lorem ut libero malesuada feugiat. curabitur non nulla  sit amet nisl tempus convallis quis ac lectus. curabitur arcu erat,  accumsan id imperdiet et, porttitor at sem. cras ultricies ligula sed magna dictum porta.',
          image: 'https://res.cloudinary.com/fuchodeveloper/image/upload/v1516760699/noodles_c6ltkq.jpg',
          views: 59,
          upVotes: 0,
          downVotes: 0,
          recipeOwnerView: true,
          createdAt: '2018-02-04T16:52:28.768Z',
          updatedAt: '2018-02-06T02:55:06.389Z'
        }
      ],
      pageCount: 1
    };
    shallow(<MyRecipesPage {...props} componentWillReceiveProps={spy} />);
    action.componentWillReceiveProps(nextProps);
    expect(action.state.recipes).toBe(nextProps.recipes);
    expect(action.state.pageCount).toBe(nextProps.pageCount);
  });
});
