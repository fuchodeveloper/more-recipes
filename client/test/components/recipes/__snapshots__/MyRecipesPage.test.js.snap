// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyRecipesPage component MyRecipesPage component snapshot should render with right amount of elements 1`] = `undefined`;
